lib/experimental/graph.nim

```diff
@@ -0,0 +1,428 @@
+
+func getEdge*[Id](graph: HGraph[Id], source, target: Id): HEdge[Id] =
+  if source.id in graph.structure.outgoingIndex:
+    for outEdge in graph.structure.outgoingIndex[source.id]:
+      if graph.structure.edgeMap[outEdge.id][1] == target:
+        return outEdge
+
+  raise newException(ValueError, "No edge for source/target pair")
+
+func edgeCount*[Id](graph: HGraph[Id]): int =
+  for node, targets in pairs(graph.edges):
+    result += targets.len()
+
+func nodeCount*[Id](graph: HGraph[Id]): int =
+  graph.edges.len()
+
+iterator nodes*[Id](graph: HGraph[Id]): Id =
+  for i, _ in pairs(graph.edges):
+    yield i
+
+proc allNodes*[Id](graph: HGraph[Id]): seq[Id] =
+  for node in nodes(graph):
+    result.add node
+
+iterator edgePairs*[Id](graph: HGraph[Id]): HEdge[Id] =
+  for node, targets in pairs(graph.edges):
+    for target in targets:
+      yield (node, target)
+
+func computeInDegrees*[Id](graph: HGraph[Id]): CountTable[Id] =
+  for (src, dst) in edgePairs(graph):
+    result.inc dst
+
+func inDeg*[Id](graph: HGraph[Id], node: Id): int {.inline.} =
+  ## Return in degree for @arg{node}
+  return computeInDegrees(graph)[node]
+
+func outDeg*[Id](graph: HGraph[Id], node: Id): int =
+  if node in graph.edges:
+    return graph.edges[node].len()
+
+func newId[Id](graph: var HGraph[Id]): int {.inline.} =
+  result = graph.structure.maxId
+  inc graph.structure.maxId
+
+func isDirected*[Id](graph: HGraph[Id]): bool {.inline.} =
+  gpDirected in graph.structure.properties
+
+
+proc removeNode*[Id](graph: var HGraph[Id], node: Id) =
+  ## Remove `node` from graph.
+  ##
+  ## Note that related edges will be removed too, if they completely loose
+  ## all related nodes. So calling `removeId("a -> b", "b")` will not only
+  ## remove node `"b"`, but also an edge.
+  for targets in mitems(graph.edges[node.id]):
+    targets.excl node.id
+
+  graph.edges.del node.id
+
+
+proc removeEdge*[Id](graph: var HGraph[Id], edge: HEdge[Id]) =
+  ## Remove `edge` from graph. Note that starting/ending nodes will not be
+  ## removed.
+  if edge.src in graph.edges:
+    graph.edges[edge.src].del graph.edges[edge.src].find(edge.dst)
+  # QUESTION error on missing edge?
+
+iterator outEdges*[Id](graph: HGraph[Id], source: Id): HEdge[Id] =
+  ## Iterate over outgoing edges for `source`
+  if source in graph.edges:
+    for it in graph.edges[source]:
+      yield (source, it)
+
+iterator inEdges*[Id](graph: HGraph[Id], target: Id): HEdge[Id] =
+  ## Iterate over ingoing edges for `target`
+  for (src, dst) in graph.edgePairs():
+    if dst == target:
+      yield (src, src)
+
+proc `$`*[Id](graph: HGraph[Id]): string =
+  var count: int = 0
+  var map: seq[(string, string, string)]
+  for edge in edgePairs(graph):
+    map.add(($graph.sourceVal(edge), $graph[edge], $graph.targetVal(edge)))
+
+  map = map.sortedByIt(it[0])
+  var sourcew = 0
+  for it in map: sourcew = max(sourcew, it[0].len)
+
+  var edgew = 0
+  for it in map: edgew = max(edgew, it[1].len)
```

Comment:
QQ: would `foldl` work or does it whine about types?

`let edgew = map.foldl(max(a, b[1].len), 0)